include/counter.h

```diff
@@ -22,35 +23,182 @@
 #include <zephyr/types.h>
 #include <stddef.h>
 #include <device.h>
+#include <stdbool.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-typedef void (*counter_callback_t)(struct device *dev, void *user_data);
+struct counter_alarm_cfg;
+
+/** @brief Alarm callback
+ *
+ * @param dev     Pointer to the device structure for the driver instance.
+ * @param cfg     Original structure owning this handler.
+ * @param ticks   Counter value that triggered the callback.
+ */
+typedef void (*counter_alarm_callback_t)(struct device *dev,
+					 const struct counter_alarm_cfg *cfg,
```

Comment:
The PR has been merged however there is one more topic I would like to discuss here.

Having the `const struct counter_alarm_cfg *cfg` parameter in the alarm callback means that we are forcing user to keep around `struct counter_alarm_cfg` variable the argument points to somewhere in the static RAM. The size of the structure is 12 bytes. In case of 3 channel counter we quickly go to 36 bytes. This RAM is there for the lifetime of the program and cannot be used for anything else. Worse, if some channels are used by different modules each module will have to keep its local copy or choose to share it as a global variable. This approach is very RAM inefficient and also results in a complex and opaque code.

The only data present in the struct that we want to keep around is pointer to a callback associated with the given channel. After call to counter_set_ch_alarm() all other fields can be thrown away (if the application wants to keep this data it can always do so). It would be more efficient if it was the driver keeping this association in a RAM table. The size of RAM used in this case would be 4 bytes * number of channels supported by the counter. Such table is already present in the drivers but it holds pointer to `struct counter_alarm_cfg` associated with the channel.

If we get rid of `const struct counter_alarm_cfg *cfg` from the callback signature user will be free to use an automatic variable when calling counter_set_ch_alarm() function -> smaller RAM footprint, simpler code, less bugs.

Reply:
isn't the idea that struct counter_alarm_cfg would be const all the way, so it would stand in a section in ROM?

Reply:
The idea behind passing adding `cfg` to the callback was to proposed as an alternative to have `user_data` there. Initial proposal was to have `user_data` provided in `set_alarm`. However, in other drivers this approach has already been used assuming that `counter_alarm_cfg` will be part of some user structure and using `CONTAINER_OF` user structure address can be retrieved.

If we change the approach to keep user callback and user_data inside the driver in RAM then we have 8 bytes per channel instead of 4 bytes per channel as it is now (pointer to cfg struct per channel). The if `struct counter_alarm_cfg` is in ROM, then current approach takes less RAM. If it is in RAM then there is a waste.

Reply:
No, it's `struct counter_config_info` structure which is constant as it is defined at compile time. `struct counter_alarm_cfg` contains run time parameters such as ticks value when the alarm will be triggered.

Reply:
@nordic-krch `struct counter_alarm_cfg` is a run time variable. Even in a very unlikely scenario when user knows during compile time all the different alarm values they want to use in their application the current approach is still inefficient. I.e. to store 5 different alarm values one would need to use 5 * 12 bytes of ROM, when in fact 5 * 4 bytes would be sufficient. The difference between `sizeof(struct counter_alarm_cfg)` and `sizeof(ticks)`.

Reply:
@mnkp, I see your point. Purpose for having `cfg` in the callback is to allow user to get his own handle and initial proposal had `user_data` instead of `cfg`. `cfg` approach is used in other drivers (e.g. adc.h) thus i was convinced to change it. Maybe in case of ADC configuration is more static than in case of timer.

I am willing to change that to arbitrary `void *user_data`. However, it will require update in existing ports. It's rather not a big change.

@anangl @tbursztyka @galak @nashif @MaureenHelm what's your opinion?

Reply:
> I am willing to change that to arbitrary `void *user_data`.

Another possibility is to pass `u8_t channel_id` in the callback. That's what watchdog API is doing. In this case we don't need to double the size of the RAM table within the driver. And user is free to set up mapping between channel_id and some internal data on their own. This would happen only when actually required by the application.

Regardless of the solution chosen we will also need to modify the signature of `counter_disable_ch_alarm()` function to take `u8_t channel_id` as a parameter instead of `const struct counter_alarm_cfg *alarm_cfg`.

Reply:
I like the approach with `u8_t channel_id` parameter proposed by @mnkp. As a lightweight one and consistent with the watchdog driver.
I think that if we decide to go this way, the `channel_id` field should be taken out of the structure and used as the second parameter in calls to `counter_set_ch_alarm`, `counter_disable_ch_alarm`, and to the callback handler.

Reply:
> However, with only `u8_t channel_id` i'm still missing way to get instance handle and i need it if module that is using the counter is multi-instance.

As @mnkp mentioned above:
> And user is free to set up mapping between channel_id and some internal data on their own. This would happen only when actually required by the application.

Do you think such mapping would be too complicated? It seems to be doable with just an array indexed with `channel_id`. Or am I missing something?

Reply:
so i'm thinking about the case where i would use counter instance for creating software pwm. one counter - one pwm instance. Instance is using all channels. When callback comes `chan_id` is not enough.

Reply:
I think that `struct device *dev` is redundant in the callback. It can be provided through `user_data` if single instance case applies. In case of multi-instance module that is using the counter `user_data` will most likely hold handle to module instance and this instance must have `dev` in it. So I would propose following modification:
- `callback(chan_id, ticks, user_data)`
- `set_alarm(dev, chan_id, cfg)`
- `disable_alarm(dev, chan_id)`

If we want to be consistent with other drivers we may keep `dev` in callback as first argument but imo it is redundant. See #8951 where arbitrary user data was needed for uart.

Reply:
I think for semi-consistency with other APIs (e.g., adc, gpio, watchdog) we should keep the first `dev` argument. I don't quite understand why that plus a channel id isn't enough for the case that you have 5 counter instances each with the same number of channels and one callback. Why can't the counter callback prototype be similar to the watchdog callback?

How about `callback(dev, chan_id, ticks)` ?

Reply:
The purpose of having such generic pointer instead of the `dev` one is to allow convenient usage of a counter device in some more complex multi-instance module or a higher-level wrapper (a similar case for UART was addressed in #8951, its description explains it nicely). @nordic-krch [mentioned](https://github.com/zephyrproject-rtos/zephyr/pull/8340#discussion_r234903819)
a software PWM as an example (one PWM uses one counter, all of its channels). For each instance of such PWM you'll probably need some structure with its data, like:
```
struct my_sw_pwm {
    struct device *counter;
    u16_t period;
    ...
```
And you'd rather want to have a common counter callback implementation for all instances. Then, having only the counter `dev` pointer passed to this callback, it would be troublesome to get the pointer to the PWM instance structure that this callback execution relates to and perform adequate actions for this instance (you'll need to do some costly mapping between the two pointers or use separate callbacks). Having a generic `user_data` as a parameter for the callback allows you to pass the proper instance pointer directly when setting up the counter.
And in case you only need the pointer to the counter device in the callback, you can still pass it through `user_data` (and this possibility makes the `dev` pointer redundant).

Reply:
@MaureenHelm actually there is no full consistency. `dma.h` and `ipm.h` are not using `dev`. In `uart.h` callback was extended to support providing `user_data` instead of `dev` and in new uart API discussion I'm pushing for removing `dev` in favor of `user_data` which basically gives more flexibility.

In general, if there is a peripheral that is in nature single instance (e.g. clock, watchdog, adc?) there will be no use case where multi-instance module will be the user of given peripheral and in that case `user_data` is not needed and `dev` might be enough but in case of peripherals which has multiple instances `user_data` is a must and there `dev` is redundant.

Reply:
For anybody trying to reconstruct design rationale see additional discussion at #11976.